ynput/ayon-core
# Addons: Preparation steps for child processes

## Changelog Description
Give addons option to initialize their environment before a pipeline process starts.

## Additional info
The preparation can be used to make sure user is logged into some service and has set environment variables that are necessary. It is achived using simple mechanism by implementing `ensure_is_process_ready` method which expects `ProcessContext` as an argument.

Right now process context contains information if is headless, which addon name and version triggered the request and which project name is used. All of the mentioned information are optional (maybe addon name and version should be required?). Using object gives option to improve what can be specified (we might also add version to `ProcessContext` for forwards/backwards compatibility?).

All addon has to do is to call `ensure_addons_are_process_ready` with object of `ProcessContext` to trigger the preparation. Addons can raise `ProcessPreparationError` with human readable message, in that case user does not have option to see/copy detail (detail contains stdout/stderr with traceback).

## Testing notes:
At this moment the functionality is implemented only in ftrack addon and used by applications addon. For that is also required to use different AYON launcher.
- AYON launcher with https://github.com/ynput/ayon-launcher/pull/112
- AYON launcher with https://github.com/ynput/ayon-launcher/pull/135 (PYTHONPATH issues)
- AYON ftrack https://github.com/ynput/ayon-ftrack/pull/109
- AYON applications https://github.com/ynput/ayon-applications/pull/1
- AYON server at least 1.3.0
1. Make sure to create packages of all of the addons.
2. It is necessary to build and install AYON launcher with the PR to register scheme uri handler.
3. Launch tray and logout from ftrack. You can close tray afterwards, but keeping it open will also test ftrack PR.
4. Start application from AYON web UI.
5. It should ask you to login to ftrack, if that fails it stops launching, otherwise it should launch the application. If you kept tray running, ftrack should update credentials.
